### PR TITLE
#158559265 Create signup page

### PR DESCRIPTION
**Pull request by chadwalt**
#### What does this PR do?
This PR creates a signup page where a user can create an account and start using the application.

#### Description of the task to be completed?
Before, there was no provision for a user to create an account with the application. This PR adds the functionality whereby the user can create an account and he/she can start using the application.

#### How should this be manually tested?
This can be tested by navigating to `http://127.0.0.1:9393/signup/`
#### What are the relevant pivotal tracker stories?
(158559265)[https://www.pivotaltracker.com/story/show/158559265]
#### Screenshots
screen shot 2018-07-10 at 19 46 22

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -42,7 +42,7 @@ build-iPhoneSimulator/
 
 # for a library or gem, you might want to ignore these files since the code is
 # intended to run in multiple environments; otherwise, check them in:
-# Gemfile.lock
+Gemfile.lock
```
Comment (igbanam) [inline .gitignore]: You don't want to ignore the `Gemfile.lock`

This guarantees that anytime this repo is run, it's run with the same [version of] gems.

It's the same argument for checking in the `db/schema.rb`

File: app/controllers/application_controller.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    set :views, 'app/views'
+    set :sessions, true
+    set :session_secret, ENV.fetch('SESSION_SECRET')
+  end
+
+  before do
+    headers "Content-Type" => "text/html; charset=utf-8"
+  end
+
+end
```
Comment (igbanam) [inline app/controllers/application_controller.rb]: The Unix standard _methinks_ requires that there's a new line at the end of every file. This is what GitHub is looking for and doesn't see that it puts this weird icon at the end of your file.

File: app/controllers/signup_controller.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    haml :signup
+  end
+
+  post "/user_info" do
+    first_name = params.fetch('first_name')
+    last_name = params.fetch('last_name')
+    username = params.fetch('username')
+    password = params.fetch('password')
+    confirm_password = params.fetch('confirm_password')
+    email = params.fetch('email')
```
Comment (igbanam) [inline app/controllers/signup_controller.rb]: So… I am not too conversant with Sinatra. I will comment on this from a Ruby perspective. 

I think this is too verbose for Ruby. Ruby is most times terse and fluent. 

Consider refactoring this with some `user_params` method which aggregates this for you?

File: app/controllers/signup_controller.rb [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+    if password != confirm_password then
+      session[:error] = "Password's didn't match"
+      redirect to("/")
+    end
+
+    session.delete(:error) if session[:error]
+
+    User.create(
```
Comment (igbanam) [inline app/controllers/signup_controller.rb]: So `User.create(user_params)` would read a bit better here

File: app/views/layout.haml [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    %meta{:charset =>"utf-8"}
+    %meta{:charset =>"utf-8", :name => "viewport", :content => "width=device-width, initial-scale=1, shrink-to-fit=no"}
+
+    %link{:rel => "stylesheet",
+          :href => "https://stackpath.bootstrapcdn.com/bootstrap/4.1.1/css/bootstrap.min.css",
+          :integrity => "sha384-WskhaSGFgHYWDcbwN70/dfYBj47jz9qbsMId/iRN3ewGhXQFZCSftd1LZCfmhktB",
+          :crossorigin =>"anonymous"}
+    %link{:rel => "stylesheet", :href => "css/#{@css_link}"}
+
+    %title Yiigaa
```
Comment (igbanam) [inline app/views/layout.haml:13]: If this title is used here like this, is it worthwhile declaring it in the `ApplicationController`?

File: app/views/signup.haml [hunk 6]
```diff
@@ -0,0 +1,32 @@
+%div{:class => ["border", "login-div"]}
+  %span{:class => "login-title"} Account Signup
+  %form{:action => "/signup/user_info", :method => "POST"}
```
Comment (igbanam) [inline app/views/signup.haml:3]: can the simple_form gem help make this less work ?

Comment (igbanam) [reply app/views/signup.haml:3]: If you guys are not allowed to use gems, that's fine

File: config/mongoid.yml [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+      hosts:
+        - localhost:27017
+
+production:
+  clients:
+    default:
+      database: micro_learning_db
+      hosts:
+        - localhost:27017
+
```
Comment (igbanam) [inline config/mongoid.yml]: The new line I spoke about in an earlier comment should be empty. This one has some space in it, so it's not considered as a new line at EOF

File: spec/controllers/signup_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,11 @@
+require_relative '../spec_helper'
+require_relative '../../app/controllers/signup_controller'
+require_relative '../../app/models/user_model'
+
+describe SignupController do
+  it "should display the signup page" do
+    get "/"
+      expect(last_response.status).to eq 200
```
Comment (igbanam) [inline spec/controllers/signup_controller_spec.rb]: only have one expectation per `it` block.

I don't think this expectation mirrors the intent of the test's description; the other one does.

File: spec/spec_helper.rb [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+ENV['RACK_ENV'] = "test"
+
+module RSpecMixin
+  include Rack::Test::Methods
+  def app
+    described_class
+  end
+end
+
+RSpec.configure {|config| config.include RSpecMixin}
```
Comment (igbanam) [inline spec/spec_helper.rb]: We need an empty new line here

**Review by igbanam — COMMENTED** [review]
I am not that conversant with Sinatra, so most of these comments are ruby-esque

File: .gitignore (hunk 1 above)
Comment (chadwalt) [reply .gitignore]: Understood, let me put it back.

File: app/views/layout.haml (hunk 5 above)
Comment (chadwalt) [reply app/views/layout.haml:13]: This has been addressed

File: app/views/signup.haml (hunk 6 above)
Comment (chadwalt) [reply app/views/signup.haml:3]: cool

File: app/controllers/signup_controller.rb [hunk 10]
```diff
@@ -0,0 +1,30 @@
+require_relative './application_controller'
+require_relative '../models/user'
+
+# Create account for a new user
+class SignupController < ApplicationController
+  error Mongo::Error::OperationFailure do |error|
+    if error.message.include? 'E11000'
+      session[:error] = 'Email taken, place use another'
```
Comment (cmwendwa) [inline app/controllers/signup_controller.rb]: Think you meant `please`, right?

**Review by cmwendwa — APPROVED** [review]
Looking good apart from the single potential typo comment.

File: app/controllers/signup_controller.rb (hunk 10 above)
Comment (chadwalt) [reply app/controllers/signup_controller.rb]: addressed

**Pull request merged**